acceptance/tests/connect/connect_inject_test.go

```diff
@@ -16,25 +18,147 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// Test that Connect works in a default and a secure installation.
+// TestConnectInject tests that Connect works in a default and a secure installation.
 func TestConnectInject(t *testing.T) {
-	cases := []struct {
+	cases := map[string]struct {
+		clusterGen  func(*testing.T, map[string]string, environment.TestContext, *config.TestConfig, string) consul.Cluster
+		releaseName string
 		secure      bool
 		autoEncrypt bool
 	}{
-		{false, false},
-		{true, false},
-		{true, true},
+		"Helm install without secure or auto-encrypt": {
+			clusterGen:  consul.NewHelmCluster,
+			releaseName: helpers.RandomName(),
+		},
+		"Helm install with secure": {
+			clusterGen:  consul.NewHelmCluster,
+			releaseName: helpers.RandomName(),
+			secure:      true,
+		},
+		"Helm install with secure and auto-encrypt": {
+			clusterGen:  consul.NewHelmCluster,
+			releaseName: helpers.RandomName(),
+			secure:      true,
+			autoEncrypt: true,
+		},
+		"CLI install without secure or auto-encrypt": {
+			clusterGen:  consul.NewCLICluster,
+			releaseName: consul.CLIReleaseName,
+		},
+		"CLI install with secure": {
+			clusterGen:  consul.NewCLICluster,
+			releaseName: consul.CLIReleaseName,
+			secure:      true,
+		},
+		"CLI install with secure and auto-encrypt": {
+			clusterGen:  consul.NewCLICluster,
+			releaseName: consul.CLIReleaseName,
+			secure:      true,
+			autoEncrypt: true,
+		},
 	}
 
-	for _, c := range cases {
-		name := fmt.Sprintf("secure: %t; auto-encrypt: %t", c.secure, c.autoEncrypt)
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			cfg := suite.Config()
+			ctx := suite.Environment().DefaultContext(t)
+
+			connHelper := ConnectHelper{
+				ClusterGenerator: c.clusterGen,
+				Secure:           c.secure,
+				AutoEncrypt:      c.autoEncrypt,
+				ReleaseName:      c.releaseName,
+				T:                t,
+				Ctx:              ctx,
+				Cfg:              cfg,
+			}
+
+			connHelper.Install()
+			connHelper.DeployClientAndServer()
+			connHelper.TestConnectionFailureWithoutIntention()
+			connHelper.CreateIntention()
+			connHelper.TestConnectionSuccess()
+			connHelper.TestConnectionFailureWhenUnhealthy()
+		})
+	}
+}
+
+// TestConnectInjectOnUpgrade tests that Connect works before and after an upgrade is performed on the cluster.
+func TestConnectInjectOnUpgrade(t *testing.T) {
+	cases := map[string]struct {
+		clusterGen          func(*testing.T, map[string]string, environment.TestContext, *config.TestConfig, string) consul.Cluster
+		releaseName         string
+		secure, autoEncrypt bool
+		initial, upgrade    map[string]string
+	}{
+		"CLI upgrade changes nothing": {
+			clusterGen:  consul.NewCLICluster,
+			releaseName: consul.CLIReleaseName,
+		},
+		"CLI upgrade from 1.10 to 1.11": {
+			clusterGen:  consul.NewCLICluster,
+			releaseName: consul.CLIReleaseName,
+			initial: map[string]string{
+				"global.image": "hashicorp/consul:1.10.7",
```

Comment:
I think we need to not hardcode these versions-- you could automatically fetch the two most recent consul versions to perform the upgrade. The versions could be fetched in a step of the pipeline using the github api, and passed into the acceptance tests as flags.

Another option we had discussed before for this test was grabbing the n-1 version of the CLI itself to install with and upgrading to the latest version of the CLI. This would possibly involve a pipeline step like above, to fetch the n-1 version of the CLI. This may not even need the github api, and could parse a previous version from the changelog, and use `apt` to fetch the CLI at that version in the pipeline step. Whenever we upgrade the CLI itself, we'd be bumping the consul-k8s image as part of that upgrade so it'd trigger a decent change to test.

Reply:
I added a way to fetch the versions so I can make this test dynamic. I also will need to write that test that uses 1 version of the CLI then another to upgrade. I think that is a separate PR. I'll make a task for it.

Reply:
awesome thanks for working on this! I think the upgrade case with CLI versions could be a part of this PR since it's an upgrade smoke test case, but if not, I think it'd be the highest priority subtask within upgrade tests since I'd consider the upgrade smoke tests task to encompass both these cases.

Reply:
Should I have it exercise the same connect inject behavior as this collection of tests?

Reply:
I'll close this conversation and add a note to do the upgrade of the CLI itself.

Reply:
Given eng sync today, I think it makes sense to not include upgrading the CLI itself (bc it could result in a consul image change). I thought eventually clients and servers could come up with a one-shot upgrade but it seems thats not the case! Sorry about that!